src/mlpack/methods/ann/layer/layer_norm_impl.hpp

```diff
@@ -0,0 +1,123 @@
+/**
+ * @file layer_norm_impl.hpp
+ * @author Shikhar Jaiswal
+ *
+ * Implementation of the Layer Normalization class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_LAYERNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_LAYERNORM_IMPL_HPP
+
+// In case it is not included.
+#include "layer_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+LayerNorm<InputDataType, OutputDataType>::LayerNorm() :
+    size(10),
+    eps(1e-8)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+LayerNorm<InputDataType, OutputDataType>::LayerNorm(
+    const size_t size, const double eps) :
+    size(size),
+    eps(eps)
+{
+  weights.set_size(size + size, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void LayerNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+  gamma.fill(1.0);
+  beta.fill(0.0);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void LayerNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  mean = arma::mean(input, 1);
+  variance = arma::var(input, 1, 1);
+
+  // Normalize the input.
+  output = input.each_col() - mean;
+  output.each_col() /= arma::sqrt(variance + eps);
+
+  // Reused in the backward and gradient step.
+  normalized = output;
+
+  // Scale and shift the output.
+  output.each_col() %= gamma;
+  output.each_col() += beta;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void LayerNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  const arma::mat inputMean = input.each_col() - mean;
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // Step 1: dl / dxhat *
```

Comment:
Yeah, I kind of had that doubt as well. I'll get back to you regarding this.

Reply:
@rcurtin What is your take on this? `Layer Normalization` follows directly from `Batch Normalization` (with `Deterministic` set to `false` at all times), so shouldn't the procedure for calculating `Gradients` also be the same?

Reply:
@zoq I couldn't find any definitive resource on this. The output matches the expected one though.

Reply:
They are the same, just wanted to avoid the confusion that step #number doesn't match with the number in the paper, perhaps removing the number from the step is all we need.

Reply:
Hmm, updated it. Now it's ready for merging.